[BLOG] add SciPy array API post

lucascolley opened the pull request:
Blog Post from Lucas' Open Source Internship, 2023

Still a work in progress!

## Text styling

- [x] The blog is written with [plain language](https://www.plainlanguage.gov/guidelines/) (where relevant).
- [x] If there are headers, they use the proper header tags in order to do so (with only one level-one header).
- [x] All links describe where they link to (for example, check the [Quansight labs website](https://labs.quansight.org/)).
- [x] Any kind of styling that the author uses (for example, bold for emphasis) is consistent throughout the blog.

## Non-text contents

- [x] Blog post featured image is in PNG or JPEG format, **not** SVG.
- [x] All content is represented as text (for example, images need alt text and videos need captions or descriptive transcripts).
- [x] If there are emojis, there are not more than three in a row.
- [x] Don't use [flashing gifs or videos](https://www.w3.org/TR/UNDERSTANDING-WCAG20/seizure-does-not-violate.html).
- [x] If it were to be read as plain text, the blog still makes sense and no information is missing.

lucascolley commented:
@tupui @izaid ready for review now :)

lucascolley commented on apps/labs/posts/scipy-array-api.md:
````diff
@@ -0,0 +1,436 @@
++     xp_assert_close(fft_res, fftpack_res)
+```
+
+## Machinery, Helpers and Utilities
````
I planned this to be the heading for this section, although I found there wasn't too much to say under it. Maybe changing it to just 'Machinery' or something else would be better? I ended up mentioning most of the other machinery above already.

pavithraes replied:
Thoughts on removing the heading altogether and updating the previous section heading to `How to Convert a Submodule: Machinery, Helpers, and Utilities`?

lucascolley replied:
combining it all into one section seems okay, although I'm not sure I like that heading...

rgommers replied:
Maybe it works to make it a subheading (`###`) and name it something like "Enabling and testing with non-NumPy arrays"?

lucascolley replied:
Sounds good, I've done exactly that in https://github.com/Quansight/Quansight-website/pull/770/commits/c359314a43da5b53526b6c237d02e1251c4f9eac.

asmeurer commented:
I would suggest using normal inline links instead of footnotes for links to external sites.

lucascolley commented:
> I would suggest using normal inline links instead of footnotes for links to external sites.

Addressed https://github.com/Quansight/Quansight-website/pull/770/commits/57766b695b86829419bbe8554bde25ba92ba2123.

pavithraes commented:
@lucascolley - Thanks for the post! I've added you to Storyblok, would you mind pushing an empty commit to re-run the preview deployment?

lucascolley commented:
@pavithraes preview deployment has worked :tada:

lucascolley commented on apps/labs/public/posts/scipy-array-api/scipy-array-api-hero.png:
Sorry, this has come out a bit blurry. I think the illustration itself looks fine in the other file, so I've just hacked this hero together badly. Where can I find the source file for the background?

pavithraes replied:
I think it's this file: https://drive.google.com/drive/folders/1fBcw_FzQSLSipzQ67WPJIouHEOQ-VNep?usp=drive_link
More templates: https://drive.google.com/drive/folders/1fBcw_FzQSLSipzQ67WPJIouHEOQ-VNep?usp=drive_link

lucascolley commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,441 @@
+I will keep this blog post focused on just array-agnostic code since this work in `special` should ultimately be
+replaced by a more robust dispatching mechanism, whereas the array-agnostic code is here to stay!
+
+However, as mentioned above, `fft` is a standard extension module, so the standard can still help us with some
```
```suggestion
Despite this limitation on GPU support, as mentioned above, `fft` is a standard extension module, so the standard can still help us with some
```

lucascolley commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,441 @@
+When we call `array_namespace`, a little more is done than just returning the namespace.
+If the environment variable `SCIPY_ARRAY_API` is not set, `array_api_compat.numpy`
+(the compatible wrapper for `numpy`) is returned, regardless of the type of the input array.
+This allows us to retain all pre-array API behaviour unless the user has set the environment variable.
```
I could clarify that it does this by sending all arrays down the `is_numpy` code path.

lucascolley commented on apps/labs/posts/scipy-array-api.md:
````diff
@@ -0,0 +1,441 @@
+                       overwrite_x=overwrite_x, workers=workers, plan=plan)
+```
+
+After converting the production functions, we also need to convert the tests!
````
Maybe a little too enthusiastic about modifying tests :sweat_smile: ?

```suggestion
After converting the production functions, we also need to convert the tests.
```

izaid replied:
Agreed on the removing the !, but also agree on the test enthusiasm. :)

lucascolley commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,441 @@
+which incorporate checks for matching namespaces and dtypes.
+If any tests use results or input data from NumPy, we convert everything to the `xp` namespace before using
+these assertions.
+The assertions make use of functions from `xp` where possible which helps avoid device transfers.
```
```suggestion
The assertions make use of functions from `xp` where possible which helps to avoid device transfers.
```

lucascolley commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,441 @@
+
+The current testing infrastructure allows us to test with PyTorch (CPU & GPU) and CuPy through the command line.
+For example, `python dev.py test -b pytorch` tests with PyTorch, and `python dev.py test -b all` tests with
+NumPy, CuPy, PyTorch and `numpy.array_api`.
+[`numpy.array_api`](https://numpy.org/devdocs/reference/array_api.html) is a strict minimal implementation of the
```
Another place where the full stop between module names doesn't look great.
![image](https://github.com/Quansight/Quansight-website/assets/51488791/54fa33b9-8a3e-4439-b398-b645b368d8dd)

pavithraes replied:
I agree, though I think this one is comparatively more understandable, so I'm OK with leaving it as is.

lucascolley commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,441 @@
+## Status
+
+At the time of writing, SciPy has two of its submodules converted (in the sense I have described above):
+`cluster` and `fft`.
+`linalg` and `signal` have partial coverage in progress, and some of the compiled-code functions in `special`
```
And another full stop between module names
![image](https://github.com/Quansight/Quansight-website/assets/51488791/30551639-c5b9-4ac3-a101-8577774bab54)

pavithraes replied:
We can update the previous sentence to:

> At the time of writing, SciPy has two of its submodules, `cluster` and `fft`, converted (in the sense I have described above).

lucascolley commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,441 @@
+At the time of writing, SciPy has two of its submodules converted (in the sense I have described above):
+`cluster` and `fft`.
+`linalg` and `signal` have partial coverage in progress, and some of the compiled-code functions in `special`
+have been covered, with additional calling-out to CuPy, PyTorch and JAX, in
+[the `special` PR](https://github.com/scipy/scipy/pull/19023/), as mentioned above.
```
```suggestion
have been covered, with additional calling-out to CuPy, PyTorch and JAX, in
[the `special` PR](https://github.com/scipy/scipy/pull/19023/) mentioned above.
```

lucascolley commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,441 @@
+I have opened [the first PR for `scipy.linalg`](https://github.com/scipy/scipy/pull/19260)
+which covers the functions which are part of
+[the `linalg` standard extension module](https://data-apis.org/array-api/latest/extensions/linear_algebra_functions.html).
+In the short term, we hope to see PRs covering the rest of the submodules and continued refinement of our
+testing infrastructure.
```
Will leave it to Ralf to decide exactly how to word what we'd like to see in the short term - it's a little tricky with some maintainers onboard and others not so much.

lucascolley commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,441 @@
+Going from having never contributed to Open Source to feeling proficient as a contributor, and having
+significant contributions merged towards a bleeding-edge project, has been a challenging but very special
+experience!
+Regular meetings with my mentors and the community meant that I was always able to ask questions when I was
+confused, and learn about the project and SciPy as a whole.
```
```suggestion
Regular meetings with my mentors and the community meant that I was always able to ask questions when I was
confused, and learn about the array types project and SciPy as a whole.
```

lucascolley commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,441 @@
+Sometimes, just putting your difficulties into words can help you figure things out from a different perspective!
+
+After joining the SciPy community as part of my internship, I am keen to contribute voluntarily in the future 
+and help to advance science in this way.
```
```suggestion
and help to sustain science in this way.
```

lucascolley commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,441 @@
+[Tyler Reddy](https://github.com/tylerjereddy) and [Matt Haberland](https://github.com/mdhaber)
+who have been greatly involved in the review process of many of my PRs, and
+[Melissa Weber Mendonça](https://github.com/melissawm) who hosted many of the community meetings which I
+attended (as well as coordinating the internship program!).
```
```suggestion
attended (as well as coordinating the internship program at Quansight Labs!).
```

lucascolley commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,441 @@
+- [The array API standard](https://data-apis.org/array-api/)
+- ["Toward Array Interoperability in the Scientific Python Ecosystem" SciPy 2023 Talk](https://youtu.be/16rB-fosAWw)
+- ["SciPy array types & libraries support" RFC](https://github.com/scipy/scipy/issues/18286)
+- ["Array Libraries Interoperability" blog post](https://labs.quansight.org/blog/array-libraries-interoperability)
+- [array-api-compat](https://github.com/data-apis/array-api-compat)
+- [Tracker issue for support in SciPy](https://github.com/scipy/scipy/issues/18867)
```
```suggestion
- ["Array Libraries Interoperability" Blog Post](https://labs.quansight.org/blog/array-libraries-interoperability)
- [array-api-compat](https://github.com/data-apis/array-api-compat)
- [Tracker Issue for array API support in SciPy](https://github.com/scipy/scipy/issues/18867)
```

pavithraes commented on apps/labs/posts/scipy-array-api.md:
````diff
@@ -0,0 +1,441 @@
+from [Anirudh's blog post](https://labs.quansight.org/blog/array-libraries-interoperability)).
+For example, when trying to use `scipy.fft.fft` with a CuPy array, users would be hit with this error message:
+
+```python
+>>> import scipy
+>>> import numpy as np
+>>> scipy.fft.fft(np.exp(2j * np.pi * np.arange(8) / 8))
+array([-3.44509285e-16+1.14423775e-17j,
+       …
+       0.00000000e+00+1.22464680e-16j])
+
+>>> import cupy as cp
+>>> scipy.fft.fft(cp.exp(2j * cp.pi * cp.arange(8) / 8))
+…
+KeyError: 'ALIGNED is not defined for cupy.ndarray.flags'
+```
````
I think showing `cupy` is enough

pavithraes commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,441 @@
+This PR added a lot of machinery (which I will explore in more detail later) to enable conversion and testing,
+while keeping the new behaviour behind an experimental environment variable.
+
+My first task was to convert `scipy.fft`.
+`fft` is an
```
```suggestion
My first task was to convert the `scipy.fft` module.
`fft` is an
```

Just a suggestion - feel free to modify as you see fit :)

pavithraes commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,441 @@
+The potential performance improvements that could be won with parallel algorithms, support for distributed
+arrays and support for GPUs (and other hardware accelerators) are huge, and thus "SciPy has had 'support
+for distributed and GPU arrays' on its roadmap" for five years now.
+See [this page](https://data-apis.org/array-api/latest/use_cases.html#use-case-scipy) on the use cases of the
+array API and [this blog post](https://labs.quansight.org/blog/2021/11/pydata-extensibility-vision) by
+Ivan Yashchuk and Ralf Gommers for more information.
```
```suggestion
For more information, see [this page on the use cases of the array API](https://data-apis.org/array-api/latest/use_cases.html#use-case-scipy), and [the blog post: "A vision for extensibility to GPU & distributed support for SciPy, scikit-learn, scikit-image and beyond"](https://labs.quansight.org/blog/2021/11/pydata-extensibility-vision) by
Ivan Yashchuk and Ralf Gommers.
```

tupui commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,441 @@
+different to those which they are used to.
+For large existing projects, this makes the prospect of switching to a different provider library a very daunting one.
+
+On the developer side of things, it seems like there is a lot of wasted time in reimplementing the same functionality
+for many different provider libraries.
```
Might be helpful to add a note saying that SciPy and other such libraries are still very much developer by volunteers. This can be linked to the above note saying that it was for 5 years on the roadmap. i.e. a start of an explanation of why now.

lucascolley replied:
Something like this?
```suggestion
On the developer side of things, it seems like there is a lot of wasted time in reimplementing the same functionality
for many different provider libraries.
Developer time is very valuable since these projects are still largely maintained by volunteers (and hence large
tasks like adding GPU support can stay on the roadmap for many years), so work to reduce the overall maintenance
load is key to accelerating progress.
```

tupui commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,441 @@
+detailed plan for how to treat different array inputs, and the development strategy.
+To summarise, the scope of the work is to treat all array inputs as specified in the RFC and to convert all
+existing pure Python + NumPy code to be array API compatible.
+Out of scope is converting the C/C++/Cython/Fortran code which is used within SciPy.
```
You explain this a bit bellow (`fht` section), but here it might not be clear why. If readers do not go read on the other posts, they would not really get the subtlety.

I suggest to add a short 1-2 lines to explain this. You could also add that behind the scene SciPy provides a Python API for low level functions written in other languages. If you say that, it would probably fit better where you describe first SciPy and you would emphasis the fact that these implementations in other languages are CPU only.

lucascolley replied:
```suggestion
Out of scope is converting the C/C++/Cython/Fortran code which is used within SciPy.
Behind the scenes, SciPy provides Python APIs for low level functions which are written in other languages and for the
CPU only.
These APIs are made to work with NumPy arrays - they will definitely not work for arrays on a different device to the CPU.
For these functions which use compiled code, some dispatching mechanism...
```

lucascolley replied:
I went with "Behind the scenes, SciPy provides Python APIs for low level functions which are written in other languages and for the CPU only. These APIs are made to work with NumPy arrays - they will definitely not work for arrays on a different device to the CPU, and are not guaranteed to work with other types of array on the CPU. For these functions which use compiled code, some dispatching mechanism seems like the right solution, but no particular implementation or development strategy has been settled on yet - see the discussion of `uarray` in [Anirudh's blog post](https://labs.quansight.org/blog/array-libraries-interoperability) and [the RFC](https://github.com/scipy/scipy/issues/18286) for more information."

tupui commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,441 @@
++    xp = array_namespace(a)
++
+     # size of transform
+-    n = np.shape(a)[-1]
++    n = a.shape[-1]
```
I would remove that from the diff as it could have been written like this before too.

lucascolley replied:
I actually quite like it being there since it's representative of the kind of small changes that are needed where the standard differs from the NumPy API - but happy to remove if others think so.

tupui replied:
Ah I see, things which are not in the standard and need a conversion. Then I would highlight that (comment or in the description.)

tupui commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,441 @@
+
+After converting the production functions, we also need to convert the tests!
+
+We have the `@array_api_compatible` decorator, which parametrizes `xp` with the
```
Can we have another level?

If yes, consider adding a summary at the beginning of the section with the steps needed, then go through them like you did.

tupui commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,441 @@
+[`numpy.array_api`](https://numpy.org/devdocs/reference/array_api.html) is a strict minimal implementation of the
+standard, which allows us to make sure that our array-agnostic code is _really_ array-agnostic -
+if our tests pass for `numpy.array_api`, then they will pass for any fully compliant library.
+PyTorch GPU can be enabled by setting the environment variable `SCIPY_DEVICE=cuda`.
```
Not sure if it's worth mentioning here that we can also pass `mps` to use things like PyTorch on CPU.

At the beginning of this section (or the previous one), you could put a link to the dev doc section around adding support for Array API.

lucascolley replied:
```suggestion
Testing with PyTorch on different devices can be enabled by setting the environment variable `SCIPY_DEVICE` to
different values: `'cuda'` for Nvidia GPUs or `'mps'` for MacOS devices with the Metal framework.

For more information on the implementation details, see
[the SciPy "Support for Array API" documentation](https://scipy.github.io/devdocs/dev/api-dev/array_api.html).
```

tupui commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,441 @@
+[the `linalg` standard extension module](https://data-apis.org/array-api/latest/extensions/linear_algebra_functions.html).
+In the short term, we hope to see PRs covering the rest of the submodules and continued refinement of our
+testing infrastructure.
+Looking further ahead, once all libraries which we want to support comply with the standard, we can drop
+`array-api-compat`.
```
`Looking further ahead` to me make a stronger connection with the previous statement. i.e. still linked to SciPy although the rest of the sentence talks about the ecosystem in general. I would either rephrase the beginning of the sentence or make 2 different statements. 1 for the long term goal of SciPy (is there more to do after converting the modules) and 2 the long term goal for the whole Scientific Python ecosystem.

tupui reviewed (approved):
This is a great post @lucascolley 👏 thank you!

Again, awesome job and I am glad that you are considering to stay around in the community.

It's already in a publishable state to me. I have some very minor and optional suggestions.

izaid commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,441 @@
+[Irwin Zaid](https://github.com/izaid) in my Open Source internship at Quansight Labs.
+My project was to make progress towards SciPy's goal of adopting support for the array API standard,
+which will allow users to use SciPy functions with CuPy arrays, PyTorch Tensors and potentially
+many other array/tensor libraries in the future.
```
Nice introduction! I think it's clear, very concisely describes the goal, and references the previous work appropriately.

izaid commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,441 @@
+encompasses Open Source Python projects which are community driven and used for science
+(see [scientific-python.org](https://scientific-python.org/about/) for more information).
+
+NumPy is the n-dimensional array library which SciPy is built to work with - it is SciPy's only non-optional
```
Maybe worth referencing the NumPy paper here: https://www.nature.com/articles/s41586-020-2649-2 ?

tupui replied:
👍 also the SciPy one then

izaid commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,441 @@
+of increasing interoperability between array libraries is one with widespread support.
+There are currently many hurdles which users can run into when navigating this fractured ecosystem.
+They can find that they have to convert between different array provider libraries in their projects, in order to
+make use of the full range of tools which have been developed.
```
You could consider adding a line here saying explicitly that such conversation between array libraries is a show-stopper for many projects. Am I going to convert between NumPy and CuPy arrays often? Nope, too much of a difference hit to do CPU-GPU copying all the time. Am I going to convert between NumPy and PyTorch often? Nope, that kills autodifferentiation features (and also GPU perfomance hit like CuPy).

lucascolley replied:
```suggestion
make use of the full range of tools which have been developed.
For many projects, frequent conversion between provider libraries is a complete no-go: cross-device copying
(e.g. from GPU to CPU when converting from CuPy to NumPy) can be a huge performance hit, and 
[automatic differentation](https://en.wikipedia.org/wiki/Automatic_differentiation) features are also killed.
```

izaid commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,441 @@
+
+[The Consortium for Python Data API Standards](https://data-apis.org/) has created
+[the Python array API standard](https://data-apis.org/array-api/), which is a "specification for what it means for a
+library to have an ‘array API’" (see [Aaron's talk](https://youtu.be/16rB-fosAWw)).
```
One could consider adding a line or two on the history of the Array API standard. Others know more, but I think it was years in the making and conversation between many organisations.

lucascolley replied:
I've added "The Consortium was formed in 2020 by a group of maintainers and industry stakeholders - see [the paper accompanying Aaron's talk](https://conference.scipy.org/proceedings/scipy2023/pdfs/aaron_meurer.pdf) for more information." in https://github.com/Quansight/Quansight-website/pull/770/commits/c359314a43da5b53526b6c237d02e1251c4f9eac.

izaid commented on apps/labs/posts/scipy-array-api.md:
````diff
@@ -0,0 +1,441 @@
+x = scipy.fft.ifftn(f_ishift)
+```
+
+On my machine, this demonstrated a ~20x performance improvment over using NumPy arrays in the same functions.
````
I think you need to mention the specific details of your GPU here (Nvidia whatever). The performance improvement will depend on that, and from what I recall you don't have a particularly powerful one so others might even get more.

lucascolley commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,437 @@
+---
+title: 'Array API Standard in SciPy'
```
Do we want 'Python' in the title?
```suggestion
title: 'The Python Array API Standard in SciPy'
```

tupui replied:
I prefer without personally

izaid replied:
I agree also without Python.

lucascolley commented:
The hero is still a little blurry - I think that [this Google Drawing](https://docs.google.com/drawings/d/1YXpYalvzTnV8Y8Q4rS6QQsOdwRt1Dq46JmG2N04iSQM/edit?usp=drivesdk) is high res, would somebody be able to convert this into a high res png for me? Or advise on how to do so?

tupui commented:
Doing `Export as PNG` is not good enough? There might be another issue on the website side itself otherwise.

Otherwise, to get the size of PNG manageable, I typically use: `pngquant --ext .png --force my_figure.png`

lucascolley commented:
> Doing Export as PNG is not good enough?

No, the quality is reduced quite a bit. The current hero looks fine to me on desktop but is noticeably blurry on mobile. [Here's the updated Google Drawing](https://docs.google.com/drawings/d/1YXpYalvzTnV8Y8Q4rS6QQsOdwRt1Dq46JmG2N04iSQM/edit?usp=drive_link) if anybody knows how to fix. `pngquant --ext .png --force my_figure.png` doesn't seem to have helped.

rgommers commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,438 @@
++
+     # size of transform
+-    n = np.shape(a)[-1]
++    n = a.shape[-1] # example of a change where the standard differs slightly from the NumPy API
```
How about something like:
```
# example of a change where the standard offers only one way of doing things, while NumPy has more ways
```
This is a good example of where the standard actually matches the canonical NumPy way, but NumPy has other (non-recommended but present) ways.

rgommers commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,438 @@
+  @pytest.mark.parametrize("type", [1, 2, 3, 4])
+  @pytest.mark.parametrize("norm", [None, 'backward', 'ortho', 'forward'])
+- def test_fftpack_equivalience(func, type, norm):
++ def test_fftpack_equivalience(func, type, norm, xp):
```
slightly embarrassing typo in the test name - but that is what it is:)

lucascolley replied:
Better to leave the typo in for honesty or change it?

rgommers commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,438 @@
+
+## Status
+
+At the time of writing, SciPy has two of its submodules, cluster and fft, converted 
```
minor: typeset `cluster` and `fft` as code?

lucascolley replied:
Ah yes, this regressed when applying a suggestion from GitHub.

rgommers commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,438 @@
+- ["SciPy array types & libraries support" RFC](https://github.com/scipy/scipy/issues/18286)
+- ["Array Libraries Interoperability" Blog Post](https://labs.quansight.org/blog/array-libraries-interoperability)
+- [array-api-compat](https://github.com/data-apis/array-api-compat)
+- [Tracker Issue for array API support in SciPy](https://github.com/scipy/scipy/issues/18867)
```
maybe add the paper here too?

rgommers reviewed (commented):
Great post Lucas! I have a number of comments on individual sentences; the structure overall seems good to me and I'm not missing anything important.

lucascolley commented on apps/labs/posts/scipy-array-api.md:
```diff
@@ -0,0 +1,438 @@
+make use of the full range of tools which have been developed.
+For users who decide to use a different provider library to take advantage of performance benefits, they can
+find themselves needing to learn new APIs to accomplish the same tasks, which are often similar but frustratingly
+different to those which they are used to.
```
```suggestion
different to those which they are used to.
There are also likely to be gaps in the functionality offered by the libraries, which can lead back to the problem
of converting between libraries.
```

lucascolley commented:
Thanks a lot to everyone who has reviewed! I have addressed all of the comments now. Ping me if there is anything else :)

As for the hero png, it may just be that the resolution of my illustration is the bottleneck.

Could we try uploading an svg? I've seen that the guidance says to use a png or jpeg, but some other blog posts have been merged with an svg.

If anyone would like to try to fix the resolution, here is a link to the drawing on my personal drive which will stay up after my internship ends: https://docs.google.com/drawings/d/1VyH4Wep7JN2tfkdzobOSnVXseWlZ7MlP3j3SE-zrvyA/edit?usp=drivesdk

tupui reviewed (approved):
Thanks again Lucas. Formal approval from my side.

rgommers reviewed (approved):
Looks great to me too, ready to go I'd say!

@pavithraes I assume that there is a schedule for merging these posts, and that won't all happen in the same week? If possible, it'd be useful to merge this post in the first batch, because once it's published it is useful to point to in ongoing PRs/conversations (there are a lot of moving parts here).

lezcano commented:
Building on what Ralf said, could we even have this post published tomorrow? Thank you!

melissawm commented on apps/labs/posts/scipy-array-api.md:
````diff
@@ -0,0 +1,470 @@
+x = scipy.fft.ifftn(f_ishift)
+```
+
+On my machine (AMD Ryzen 5 2600 & NVIDIA GeForce GTX 1060 6GB), this demonstrated a ~15x performance improvment over
````
```suggestion
On my machine (AMD Ryzen 5 2600 & NVIDIA GeForce GTX 1060 6GB), this demonstrated a ~15x performance improvement over
```

melissawm reviewed (approved):
Looks great with one tiny typo - @lucascolley if you are around would you mind fixing that before I merge? Thanks!

melissawm commented:
Thanks @lucascolley for the great content and reviewers @rgommers @tupui 🎉

lucascolley commented:
Thanks @melissawm !!

Does it matter that the published date says September 25th?

I'll also copy my comments about the blurry hero to an issue but feel free to close it if the bottleneck is the illustration.
